nemo_skills/code_execution/local_sandbox/local_sandbox_server.py

```diff
@@ -13,41 +13,97 @@
+            temp_file_name = temp_file.name
+            temp_file.write(generated_code.encode('utf-8'))
+
+        result = subprocess.run(
+            ['lake', 'env', '--dir', project_path, 'lean', temp_file_name],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            timeout=timeout,
+            cwd=project_path,  # Ensure we are in the correct working directory
+        )
+
+        if result.returncode == 0:
+            process_status = "completed"
+        else:
+            process_status = "failed"
+
+        return {
+            "process_status": process_status,
+            "stdout": result.stdout.decode('utf-8'),
+            "stderr": result.stderr.decode('utf-8'),
+        }
+
+    except subprocess.TimeoutExpired:
+        return {"process_status": "timeout", "stdout": "Timed out", "stderr": "Timed out"}
+    except Exception as e:
+        print(f"Error: {str(e)}")
+        return {"process_status": "error", "stdout": "", "stderr": str(e)}
+    finally:
+        # Safely remove the temporary file if it was created
+        if temp_file_name and os.path.exists(temp_file_name):
+            os.remove(temp_file_name)
+
+
 # need to memory-limit to avoid common errors of allocating too much
 # but this has to be done in a subprocess to not crush server itself
 def execute_code_subprocess(generated_code, queue):
-    # this can be overriden inside generated code, so it's not a guaranteed protection
     limit = 1024 * 1024 * 1024 * 10  # 10gb - somehow with a smaller limit the server dies when numpy is used
     resource.setrlimit(resource.RLIMIT_AS, (limit, limit))
     resource.setrlimit(resource.RLIMIT_DATA, (limit, limit))
     resource.setrlimit(resource.RLIMIT_STACK, (limit, limit))
 
+    # this can be overriden inside generated code, so it's not a guaranteed protection
     sys.stdout = StringIO()
-    exec(generated_code, {})
-    queue.put(sys.stdout.getvalue())
+    try:
+        exec(generated_code, {})
+        queue.put(sys.stdout.getvalue())
+    except Exception as e:
+        print(f"Error: {str(e)}")
+        queue.put({"process_status": "error", "stdout": "", "stderr": str(e)})
 
 
+# Main Flask endpoint to handle execution requests
 @app.route("/execute", methods=["POST"])
 def execute():
     generated_code = request.json['generated_code']
     timeout = request.json['timeout']
-    # running in a separate process to ensure any kind of crashes are properly handled
-    queue = multiprocessing.Queue()
-    process = multiprocessing.Process(target=execute_code_subprocess, args=(generated_code, queue))
-    process.start()
-    process.join(timeout=timeout)
-    if process.is_alive():  # didn't finish successfully
-        process.kill()
-        return '{"process_status": "timeout", "stdout": "Timed out", "stderr": "Timed out"}'
-    return queue.get()
+    answer_format = request.json.get('answer_format', 'natural_language')
```

Comment:
Sorry, I think maybe I didn't explain this properly. Here we still use "language" key, which is python or lean4 as this function just executes code. It doesn't know whether we do it to check the answer or for whatever other reason. But inside the is_output_correct, we don't want to use the "language" parameter, but instead use this "answer_format", since there the fact that we check math answers through python code execution call is really just a technical implementation detail and has no real meaning other than that.

Reply:
Done